File: crates/sc-search/src/fold.rs
//! Case folding + NFC normalisation, applied **at index time** so that
//! case/normalisation variants are never stored twice (§4.1, "an additional
//! compression technique").
//!
//! The same fold is applied to the query, so the two sides always meet in the
//! same space. Non-UTF-8 byte sequences are folded ASCII-only and otherwise
//! preserved verbatim — a filename that is not valid UTF-8 must still be
//! findable (§4.2, "filenames are byte strings").

use unicode_normalization::UnicodeNormalization;

/// Fold a byte string for matching/indexing.
///
/// * valid UTF-8, pure ASCII → `to_ascii_lowercase` (no allocation beyond the
///   copy, no Unicode tables touched — the common case)
/// * valid UTF-8, non-ASCII → NFC then Unicode lowercase
/// * invalid UTF-8 → ASCII lowercase on the raw bytes, everything else kept
pub fn fold(bytes: &[u8]) -> Vec<u8> {
    match std::str::from_utf8(bytes) {
        Ok(s) if s.is_ascii() => s.to_ascii_lowercase().into_bytes(),
        Ok(s) => {
            let nfc: String = s.nfc().collect();
            nfc.to_lowercase().into_bytes()
        }
        Err(_) => bytes.to_ascii_lowercase(),
    }
}

/// [`fold`] for a `&str`.
pub fn fold_str(s: &str) -> Vec<u8> {
    fold(s.as_bytes())
}

/// True when folding would be a no-op, i.e. the bytes are already ASCII and
/// already lowercase. Lets hot loops skip the allocation.
pub fn is_folded_ascii(bytes: &[u8]) -> bool {
    bytes
        .iter()
        .all(|b| b.is_ascii() && !b.is_ascii_uppercase())
}

/// Case-insensitive ASCII substring search that allocates nothing.
///
/// `needle` must already be folded and ASCII. Used for the overwhelmingly
/// common Latin query against a Latin filename.
pub fn contains_ascii_ci(haystack: &[u8], needle: &[u8]) -> bool {
    if needle.is_empty() {
        return true;
    }
    if haystack.len() < needle.len() {
        return false;
    }
    let first = needle[0];
    for i in 0..=(haystack.len() - needle.len()) {
        if haystack[i].to_ascii_lowercase() != first {
            continue;
        }
        if haystack[i..i + needle.len()]
            .iter()
            .zip(needle)
            .all(|(h, n)| h.to_ascii_lowercase() == *n)
        {
            return true;
        }
    }
    false
}

/// Byte-exact substring search. Both sides are expected to be pre-folded.
pub fn contains(haystack: &[u8], needle: &[u8]) -> bool {
    if needle.is_empty() {
        return true;
    }
    if haystack.len() < needle.len() {
        return false;
    }
    haystack
        .windows(needle.len())
        .any(|w| w == needle)
}

/// Prefix test on pre-folded bytes.
pub fn starts_with(haystack: &[u8], needle: &[u8]) -> bool {
    haystack.len() >= needle.len() && &haystack[..needle.len()] == needle
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ascii_fold() {
        assert_eq!(fold(b"IMG_0001.JPG"), b"img_0001.jpg".to_vec());
    }

    #[test]
    fn cjk_survives_fold() {
        let s = "여름휴가사진.jpg";
        let f = fold_str(s);
        // Hangul has no case, so folding must not perturb the bytes.
        assert!(contains(&f, "휴가".as_bytes()));
        assert!(contains(&f, "여름".as_bytes()));
        assert!(contains(&f, "사진".as_bytes()));
    }

    #[test]
    fn nfd_and_nfc_fold_together() {
        // "é" as U+00E9 vs "e" + U+0301.
        let nfc = "caf\u{00e9}.txt";
        let nfd = "cafe\u{0301}.txt";
        assert_eq!(fold_str(nfc), fold_str(nfd));
    }

    #[test]
    fn invalid_utf8_is_preserved() {
        let raw = b"ABC\xff\xfe.bin";
        let f = fold(raw);
        assert_eq!(f, b"abc\xff\xfe.bin".to_vec());
        assert!(contains(&f, b"\xff\xfe"));
    }

    #[test]
    fn ci_search() {
        assert!(contains_ascii_ci(b"Vacation_Photo.JPG", b"photo"));
        assert!(!contains_ascii_ci(b"Vacation.JPG", b"photo"));
        assert!(contains_ascii_ci(b"abc", b""));
    }
}

File: go/engine/service/search/fold.go
package search

import (
	"bytes"
	"unicode/utf8"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"
	"golang.org/x/text/unicode/norm"
)

// Case folding together with NFC normalisation, performed at index time so no
// case or normalisation variant is ever stored twice. The identical fold applies
// to the query, so both sides always meet in one space.
//
// Filenames are byte strings rather than strings. A name that is not valid UTF-8
// must remain findable, so it is folded ASCII-only and otherwise preserved
// verbatim instead of being rewritten with error runes.

// lowerFull stays language-neutral. Filenames carry no locale, and a Turkish
// locale would fold ASCII "I" into a dotless form, rendering every Latin name
// containing it unfindable.
//
//nolint:gochecknoglobals // cases.Caser is immutable and documented as concurrency-safe.
var lowerFull = cases.Lower(language.Und)

// Fold normalises a byte string for indexing and matching.
func Fold(b []byte) []byte {
	if !utf8.Valid(b) || isASCII(b) {
		// Not UTF-8, so no Unicode operation is meaningful, or plain ASCII,
		// which is the common case and the one worth keeping off the tables.
		// Either way ASCII case still folds and every other byte survives.
		return asciiLower(b)
	}
	// The full Unicode lowercase mapping, not the simple one: they differ on
	// characters whose lowercase form is longer than one rune, and a name
	// spelled either way has to fold to the same bytes or it is findable by
	// only one of its spellings.
	return []byte(lowerFull.String(norm.NFC.String(string(b))))
}

// FoldString applies Fold to a string.
func FoldString(s string) []byte { return Fold([]byte(s)) }

// IsFoldedASCII reports whether folding would leave the input unchanged, letting
// a hot loop avoid the allocation.
func IsFoldedASCII(b []byte) bool {
	for _, c := range b {
		if c >= utf8.RuneSelf || (c >= 'A' && c <= 'Z') {
			return false
		}
	}
	return true
}

func isASCII(b []byte) bool {
	for _, c := range b {
		if c >= utf8.RuneSelf {
			return false
		}
	}
	return true
}

func asciiLower(b []byte) []byte {
	out := make([]byte, len(b))
	for i, c := range b {
		out[i] = lowerByte(c)
	}
	return out
}

func lowerByte(c byte) byte {
	if c >= 'A' && c <= 'Z' {
		return c + 'a' - 'A'
	}
	return c
}

// ContainsASCIIFold performs a case-insensitive ASCII substring search without
// allocating. needle must arrive already folded and ASCII, covering the
// overwhelmingly common case of a Latin query against a Latin filename.
func ContainsASCIIFold(haystack, needle []byte) bool {
	if len(needle) == 0 {
		return true
	}
	if len(haystack) < len(needle) {
		return false
	}
	first := needle[0]
	for i := 0; i+len(needle) <= len(haystack); i++ {
		if lowerByte(haystack[i]) != first {
			continue
		}
		match := true
		for j := range needle {
			if lowerByte(haystack[i+j]) != needle[j] {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}

// Contains performs a byte-exact substring search, expecting both sides to be
// pre-folded.
func Contains(haystack, needle []byte) bool { return bytes.Contains(haystack, needle) }

// HasPrefix tests a prefix over pre-folded bytes.
func HasPrefix(haystack, needle []byte) bool { return bytes.HasPrefix(haystack, needle) }
